File: webapp/public/js/page.js
//
// page.js
//
// Javascript for page.py
// 
// Copyright (C) 2020 by G3UKB Bob Cowdery
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//    
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//    
// You should have received a copy of the GNU General Public License
//  along with this program; if not, write to the Free Software
//  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//    
//  The author can be reached by email at:   
//     bob@bobcowdery.plus.com
//

//////////////////////////////////////////////////////////////////
// Main code
$(document).ready(function() {
  
  ////////////////////////////////////////////
  // Dial jog frequency
  do_dial();
  
  ////////////////////////////////////////////
  // Scroll frequency
  do_scroll();
  
  ////////////////////////////////////////////
  // Slider frequency
  do_slider();
  
  ////////////////////////////////////////////
  // Frequency increment selection
  do_increment();
  
  ////////////////////////////////////////////
  // Mode selection
  do_mode();
  
  ////////////////////////////////////////////
  // Band selection
  do_band();
});

//////////////////////////////////////////////////////////////////
// Support functions

////////////////////////////////////////////
// Mouse over highlight rate
function highlight_rate(rate) {
    $("#100KHz").css('background', "#e5e5e5");
    $("#10KHz").css('background', "#e5e5e5");
    $("#1KHz").css('background', "#e5e5e5");
    $("#100Hz").css('background', "#e5e5e5");
    $("#10Hz").css('background', "#e5e5e5");
    rate.css('background', "#a5e5e5");
  }

////////////////////////////////////////////
// Mouse over highlight mode
function highlight_mode(mode) {
    $("#LSB").css('background', "#e5e5e5");
    $("#USB").css('background', "#e5e5e5");
    $("#AM").css('background', "#e5e5e5");
    $("#FM").css('background', "#e5e5e5");
    mode.css('background', "#a5e5e5");
  }

////////////////////////////////////////////
// Mouse over highlight band  
function highlight_band(band) {
    $("#B160").css('background', "#e5e5e5");
    $("#B80").css('background', "#e5e5e5");
    $("#B40").css('background', "#e5e5e5");
    $("#B20").css('background', "#e5e5e5");
    $("#B15").css('background', "#e5e5e5");
    $("#B10").css('background', "#e5e5e5");
    $("#B2M").css('background', "#e5e5e5");
    $("#B70CM").css('background', "#e5e5e5");
    band.css('background', "#a5e5e5");
  }

////////////////////////////////////////////
// Set new frequency
function set_freq(string) {
    $("#MHz100").text(string[0]);
    $("#MHz10").text(string[1]);
    $("#MHz1").text(string[2]);
    $("#KHz100").text(string[3]);
    $("#KHz10").text(string[4]);
    $("#KHz1").text(string[5]);
    $("#Hz100").text(string[6]);
    $("#Hz10").text(string[7]);
    $("#Hz1").text(string[8]);
  }

////////////////////////////////////////////
// Set new frequency
function slider_freq() {
  // Fine tune
  $.ajax({
     type: "PUT",
     url: "/slider_service",
     data: {slider: $( "#slider" ).slider("value")}
   })
   .done(function (string) {
     set_freq(string);
   });
}
  
//////////////////////////////////////////////////////////////////
// Execute functions
  
////////////////////////////////////////////
// Do scroll exchange
function execute_scroll(e, inc) {
  if (e.originalEvent.wheelDelta > 0 || e.originalEvent.detail < 0) {
    // Scroll up
    $.ajax({
      type: "PUT",
      url: "/scroll_service",
      data: {"scroll": inc}
    })
    .done(function (string) {
      set_freq(string);
    });
  }
  else {
    // Scroll down
    $.ajax({
      type: "PUT",
      url: "/scroll_service",
      data: {"scroll": -inc}
    })
    .done(function (string) {
      set_freq(string);
    });
  }
  e.preventDefault();
}

////////////////////////////////////////////
// Do slider frequency
function do_slider() {
  $( "#slider" ).slider({
      orientation: "horizontal",
      max: 100,
      value: 50,
      slide: function( event, ui ) {
        slider_freq();
      }
    });
}

////////////////////////////////////////////
// Do frequenct jog dial
function do_dial() {
  var el = document.getElementById('dial');
  var dial = JogDial(el, {debug: false});
  dial.on("mousemove", function(e){
      $.ajax({
          type: "PUT",
          url: "/dial_service",
          data: {"rotation": e.target.rotation}
      })
      .done(function (string) {
        set_freq(string);
      });
    e.preventDefault();
  });
}

////////////////////////////////////////////
// Do scroll frequency
function do_scroll() {
  $(Hz1).bind('mousewheel DOMMouseScroll', function(e){
    execute_scroll(e, 1);
  });
  $(Hz10).bind('mousewheel DOMMouseScroll', function(e){
    execute_scroll(e, 10);
  });
  $(Hz100).bind('mousewheel DOMMouseScroll', function(e){
    execute_scroll(e, 100);
  });
  $(KHz1).bind('mousewheel DOMMouseScroll', function(e){
    execute_scroll(e, 1000);
  });
  $(KHz10).bind('mousewheel DOMMouseScroll', function(e){
    execute_scroll(e, 10000);
  });
  $(KHz100).bind('mousewheel DOMMouseScroll', function(e){
    execute_scroll(e, 100000);
  });
  $(MHz1).bind('mousewheel DOMMouseScroll', function(e){
    execute_scroll(e, 1000000);
  });
  $(MHz10).bind('mousewheel DOMMouseScroll', function(e){
    execute_scroll(e, 10000000);
  });
  $(MHz100).bind('mousewheel DOMMouseScroll', function(e){
    execute_scroll(e, 100000000);
  });
}

////////////////////////////////////////////
// Set frequency increment
function do_increment() {
  $("#100KHz").click(function(e) {
    highlight_rate($("#100KHz"));
    $.ajax({
      type: "PUT",
      url: "/rate_service",
      data: {
              "rate": "100KHz"
          }
    });
    e.preventDefault();
  });
  $("#10KHz").click(function(e) {
    highlight_rate($("#10KHz"));
    $.ajax({
      type: "PUT",
      url: "/rate_service",
      data: {
              "rate": "10KHz"
          }
    });
    e.preventDefault();
  });
  $("#1KHz").click(function(e) {
    highlight_rate($("#1KHz"));
    $.ajax({
      type: "PUT",
      url: "/rate_service",
      data: {
              "rate": "1KHz"
          }
    });
    e.preventDefault();
  });
  $("#100Hz").click(function(e) {
    highlight_rate($("#100Hz"));
    $.ajax({
      type: "PUT",
      url: "/rate_service",
      data: {
              "rate": "100Hz"
          }
    });
    e.preventDefault();
  });
  $("#10Hz").click(function(e) {
    highlight_rate($("#10Hz"));
    $.ajax({
      type: "PUT",
      url: "/rate_service",
      data: {
              "rate": "10Hz"
          }
    });
    e.preventDefault();
  });
}

////////////////////////////////////////////
// Set mode
function do_mode() {
  $("#LSB").click(function(e) {
    highlight_mode($("#LSB"));
    $.ajax({
      type: "PUT",
      url: "/mode_service",
      data: {
              "mode": "LSB"
          }
    });
    e.preventDefault();
  });
  $("#USB").click(function(e) {
    highlight_mode($("#USB"));
    $.ajax({
      type: "PUT",
      url: "/mode_service",
      data: {
              "mode": "USB"
          }
    });
    e.preventDefault();
  });
  $("#AM").click(function(e) {
    highlight_mode($("#AM"));
    $.ajax({
      type: "PUT",
      url: "/mode_service",
      data: {
              "mode": "AM"
          }
    });
    e.preventDefault();
  });
  $("#FM").click(function(e) {
    highlight_mode($("#FM"));
    $.ajax({
      type: "PUT",
      url: "/mode_service",
      data: {
              "mode": "FM"
          }
    });
    e.preventDefault();
  });
}

////////////////////////////////////////////
// Set band
function do_band() {
  $("#B160").click(function(e) {
    highlight_band($("#B160"));
    $.ajax({
      type: "PUT",
      url: "/band_service",
      data: {
              "band": "160m"
          }
    })
      .done(function (string) {
        set_freq(string);
      });
    e.preventDefault();
  });
  $("#B80").click(function(e) {
    highlight_band($("#B80"));
    $.ajax({
      type: "PUT",
      url: "/band_service",
      data: {
              "band": "80m"
          }
    })
      .done(function (string) {
        set_freq(string);
      });
    e.preventDefault();
  });
  $("#B40").click(function(e) {
    highlight_band($("#B40"));
    $.ajax({
      type: "PUT",
      url: "/band_service",
      data: {
              "band": "40m"
          }
    })
      .done(function (string) {
        set_freq(string);
      });
    e.preventDefault();
  });
  $("#B20").click(function(e) {
    highlight_band($("#B20"));
    $.ajax({
      type: "PUT",
      url: "/band_service",
      data: {
              "band": "20m"
          }
    })
      .done(function (string) {
        set_freq(string);
      });
    e.preventDefault();
  });
  $("#B15").click(function(e) {
    highlight_band($("#B15"));
    $.ajax({
      type: "PUT",
      url: "/band_service",
      data: {
              "band": "15m"
          }
    })
      .done(function (string) {
        set_freq(string);
      });
    e.preventDefault();
  });
  $("#B10").click(function(e) {
    highlight_band($("#B10"));
    $.ajax({
      type: "PUT",
      url: "/band_service",
      data: {
              "band": "10m"
          }
    })
      .done(function (string) {
        set_freq(string);
      });
    e.preventDefault();
  });
  $("#B2M").click(function(e) {
    highlight_band($("#B2M"));
    $.ajax({
      type: "PUT",
      url: "/band_service",
      data: {
              "band": "2m"
          }
    })
      .done(function (string) {
        set_freq(string);
      });
    e.preventDefault();
  });
  $("#B70CM").click(function(e) {
    highlight_band($("#B70CM"));
    $.ajax({
      type: "PUT",
      url: "/band_service",
      data: {
              "band": "70cm"
          }
    })
      .done(function (string) {
        set_freq(string);
      });
    e.preventDefault();
  });
}

File: webapp/page.py
#!/usr/bin/env python
#
# page.py
#
# The one and only HTML page
# 
# Copyright (C) 2020 by G3UKB Bob Cowdery
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#    
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#    
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#    
#  The author can be reached by email at:   
#     bob@bobcowdery.plus.com
#

# System imports
import os, sys
import pickle

# Library imports

# Application imports

#==============================================================================================
# PUBLIC
#==============================================================================================
 
#-------------------------------------------------
# Get the one and only page     
def get_page(name, model):
    
    index_html = '''
    <html>
    <head>
        <link href="/static/css/page.css" rel="stylesheet">
        <script src="http://code.jquery.com/jquery-2.0.3.min.js"></script>
        <script type="text/javascript" src="/static/js/jogDial.js"></script>
        <script type="text/javascript" src="/static/js/page.js"></script>
    </head>
    <body>
        <div id="container" class="grid-container">
            <div id="header" class="header-item"> %s </div>
            <div id="frequency" class="freq-item"> %s </div>
            <div id="dial" class="dial-item"></div>
            <div id="increment" class="inc-item"> %s </div>
            <div id="mode" class="mode-item"> %s </div>
            <div id="band" class="band-item"> %s </div>
            <div id="footer" class="footer-item"> %s </div>
        </div>
    </body>
    </html>
    ''' % (get_header(name), get_frequency(model), get_increment(model), get_mode(model), get_band(model), get_footer())
    return index_html

#==============================================================================================
# PRIVATE
#==============================================================================================
        
#-------------------------------------------------
# Header HTML
def get_header(name):     
    return "<h1>%s</h1>" % (name)

#-------------------------------------------------
# Content HTML
# Frequency div
def get_frequency(model):
    
    m = model.get_model()
    content = ''' 
    <p>
    <label id="MHz100" class="MHz">0</label>
    <label id="MHz10" class="MHz">0</label>
    <label id="MHz1" class="MHz">0</label>
    <label id="Sep1" class="Sep">.</label>
    <label id="KHz100" class="KHz">0</label>
    <label id="KHz10" class="KHz">0</label>
    <label id="KHz1" class="KHz">0</label>
    <label id="Sep2" class="Sep">.</label>
    <label id="Hz100" class="Hz">0</label>
    <label id="Hz10" class="Hz">0</label>
    <label id="Hz1" class="Hz">0</label>
    </p>
    '''
    return content

#------------------------------------------------- 
# Increament div
def get_increment(model):
    
    m = model.get_model()
    content = ''' 
    <p>
    <label id="100KHz" class="IncKHz">100KHz</label>
    <label id="10KHz" class="IncKHz">10KHz</label>
    <label id="1KHz" class="IncKHz">1KHz</label>
    <label id="100Hz" class="IncHz">100Hz</label>
    <label id="10Hz" class="IncHz">10Hz</label>
    </p>
    '''
    return content

#------------------------------------------------- 
# Mode div
def get_mode(model):
    
    m = model.get_model()
    content = ''' 
    <p>
    <label id="LSB" class="Mode">LSB</label>
    <label id="USB" class="Mode">USB</label>
    <label id="AM" class="Mode">AM</label>
    <label id="FM" class="Mode">FM</label>
    </p>
    '''
    return content

#------------------------------------------------- 
# Band div
def get_band(model):
    
    m = model.get_model()
    content = ''' 
    <p>
    <label id="B160" class="Band">160</label>
    <label id="B80" class="Band">80</label>
    <label id="B40" class="Band">40</label>
    <label id="B20" class="Band">20</label>
    <label id="B15" class="Band">15</label>
    <label id="B10" class="Band">10</label>
    <label id="B2M" class="Band">2M</label>
    <label id="B70CM" class="Band">70CM</label>
    </p>
    '''
    return content
    
#-------------------------------------------------
# Footer HTML
def get_footer():
    return "Bob Cowdery - G3UKB"
